PKPaymentMethod fails strict mode decoding when using private label cards

aprotyas opened the pull request:
#### 96136876b937be459c13e69c62bbc411457bf739
<pre>
PKPaymentMethod fails strict mode decoding when using private label cards
<a href="https://bugs.webkit.org/show_bug.cgi?id=267480">https://bugs.webkit.org/show_bug.cgi?id=267480</a>
<a href="https://rdar.apple.com/120005200">rdar://120005200</a>

Reviewed by Alex Christensen.

When using private label cards, the PKPaymentMethod object contains
non-null PKPaymentPass data, which leads to an NSKU decode failure when
sending the WebPaymentCoordinator::DidSelectPaymentMethod IPC message
because PKPaymentPass is not strict decodable.

This patch lets us fall back to secure coding for the PKPaymentMethod
type on configurations where the PKPaymentPass type is not strict
decodable, such as on macOS downlevels, which are not privy to upstream
PassKit changes that address the underlying issue.

* Source/WTF/wtf/PlatformHave.h:
Add a new compile-time PlatformHave flag that lets us opt out of
strictly decoding PKPaymentMethod, instead falling back to secure
decoding, for platform configurations that cannot perform strict
decoding of the PKPaymentPass type.

* Source/WebKit/Shared/Cocoa/ArgumentCodersCocoa.mm:
(IPC::shouldEnableStrictMode):

Canonical link: <a href="https://commits.webkit.org/273108@main">https://commits.webkit.org/273108@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/510b6e10d207ff1f5e51bdbc9666d94cd2fa010f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/34274 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/13083 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/36266 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/36958 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/31042 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/15469 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/10208 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/30059 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/34786 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/11077 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/30577 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/9668 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/9771 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/30625 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/38248 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/29173 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/31141 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/30942 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/35851 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/34191 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/9898 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/7765 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/33776 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/11692 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/40737 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/7889 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/10455 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/8485 "Passed tests") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/10725 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebKit/Shared/Cocoa/ArgumentCodersCocoa.mm:
```diff
@@ -477,6 +478,7 @@ static bool shouldEnableStrictMode(Decoder& decoder, const HashSet<Class>& allow
             || messageName == IPC::MessageName::AuthenticationManager_CompleteAuthenticationChallenge; // UIP -> NP
     };
 
+#if ENABLE(DATA_DETECTION) || ENABLE(REVEAL)
```
I don't understand why this is needed now but not before.

aprotyas replied:
To fix the watchOS build. The "newer OS-es shortcut" at the top of `shouldEnableStrictMode` is now evaluating to false for watchOS 10, while we're using `HAVE(SECURE_ACTION_CONTEXT)` outside of `ENABLE(DATA_DETECTION)`, which broke the initial assumption under which that code was introduced...

```cpp
#if ((PLATFORM(MAC) && __MAC_OS_X_VERSION_MIN_REQUIRED >= 140000) \
    || ((PLATFORM(IOS) || PLATFORM(MACCATALYST)) && __IPHONE_OS_VERSION_MIN_REQUIRED >= 170000)) \
    || PLATFORM(VISION)
// Only used inside ENABLE(DATA_DETECTION)
#define HAVE_SECURE_ACTION_CONTEXT 1
#endif
```

aprotyas committed:
PKPaymentMethod fails strict mode decoding when using private label cards
https://bugs.webkit.org/show_bug.cgi?id=267480
rdar://120005200

Reviewed by Alex Christensen.

When using private label cards, the PKPaymentMethod object contains
non-null PKPaymentPass data, which leads to an NSKU decode failure when
sending the WebPaymentCoordinator::DidSelectPaymentMethod IPC message
because PKPaymentPass is not strict decodable.

This patch lets us fall back to secure coding for the PKPaymentMethod
type on configurations where the PKPaymentPass type is not strict
decodable, such as on macOS downlevels, which are not privy to upstream
PassKit changes that address the underlying issue.

* Source/WTF/wtf/PlatformHave.h:
Add a new compile-time PlatformHave flag that lets us opt out of
strictly decoding PKPaymentMethod, instead falling back to secure
decoding, for platform configurations that cannot perform strict
decoding of the PKPaymentPass type.

* Source/WebKit/Shared/Cocoa/ArgumentCodersCocoa.mm:
(IPC::shouldEnableStrictMode):

Canonical link: https://commits.webkit.org/273108@main
